src/config.rs

```diff
@@ -1,24 +1,54 @@
 use std::fs;
-use std::path::PathBuf;
+use std::path::{Path, PathBuf};
 
 use anyhow::Result;
 use std::io;
 
-use crate::path::toolchains_dir;
+use crate::file::{read_file, write_file};
+use crate::path::{ensure_dir_exists, hashes_dir, toolchains_dir};
 use crate::target_triple::TargetTriple;
-use crate::toolchain::RESERVED_TOOLCHAIN_NAMES;
+use crate::toolchain::{OfficialToolchainDescription, RESERVED_TOOLCHAIN_NAMES};
 
 pub struct Config {
     toolchains_dir: PathBuf,
+    hashes_dir: PathBuf,
 }
 
 impl Config {
     pub(crate) fn from_env() -> Result<Self> {
         Ok(Self {
             toolchains_dir: toolchains_dir(),
+            hashes_dir: hashes_dir(),
         })
     }
 
+    pub(crate) fn hashes_dir(&self) -> &Path {
+        self.hashes_dir.as_path()
+    }
+
+    pub(crate) fn hash_matches(
+        &self,
+        description: &OfficialToolchainDescription,
+        hash: &str,
+    ) -> bool {
+        let hash_path = self.hashes_dir.join(description.to_string());
+
+        match read_file("hash", &hash_path) {
+            Ok(h) => h == hash,
+            _ => false,
```

Comment:
Rather than treating all types of I/O errors as a hash mismatch, we should only return `false` if the I/O error indicates that the file doesn't exist, or if `h != hash`, otherwise we should probably return the error.

Reply:
The initial thought was that the only time that a toolchain should skip installation was if the hashes matches and to re-install for every other case, but I guess I shouldn't be lazy and include explicitly the situations where this happens :p